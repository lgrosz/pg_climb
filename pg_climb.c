#include "pg_climb.h"

#include <assert.h>
#include <errno.h>
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

const char *grade_type_name(uint32_t type)
{
	switch (type) {
		case VERMTYPE:
			return "V-Scale";
		default:
			return "Unknown";
	}
}


Verm *verm_create(uint8_t initial_value)
{
	Verm *verm;
	uint8_t *value;

	verm = malloc(sizeof(Verm));
	value = malloc(sizeof(uint8_t));

	memcpy(value, &initial_value, sizeof(uint8_t));
	verm->value = value;
	verm->type = VERMTYPE;

	return verm;
}

void verm_free(Verm *verm)
{
	free(verm->value);
	free(verm);
}

uint8_t verm_get_value(const Verm *verm)
{
	return *verm->value;
}

void verm_set_value(Verm *verm, uint8_t value)
{
	memcpy(verm->value, &value, sizeof(uint8_t));
}

int verm_parse(Verm *verm, const char *str)
{
	int	value;

	if (str == NULL || verm == NULL)
		return 1;

	assert(verm->value);

	if (strlen(str) < 2)
		return 1;

	if (strncasecmp(str, "v", 1) != 0)
		return 1;

	errno = 0;
	value = strtol(str + 1, NULL, 10);

	if (errno == ERANGE || value < 0 || value > 255)
		return 1;

	memcpy(verm->value, &value, sizeof(uint8_t));
	return 0;
}

Verm *verm_from_string(const char *str)
{
	Verm *verm = verm_create(0);

	if (verm_parse(verm, str) != 0) {
		verm_free(verm);
		verm = NULL;
	}

	return verm;
}

char *verm_format(const Verm *verm)
{
	char	*str;

	if (verm == NULL)
		return NULL;

	str = malloc(5 * sizeof(char));
	snprintf(str, 5, "V%d", verm_get_value(verm));

	return str;
}

Grade *grade_from_string(const char *str, uint32_t type_hint)
{
	Grade	*grade;

	switch (type_hint) {
		case ANYTYPE: // try all until one hits
		case VERMTYPE:
			if ((grade = (Grade*)verm_from_string(str)) != NULL || type_hint) break;
		default:
			grade = NULL;
	}

	return grade;
}

void grade_free(Grade *grade)
{
	switch (grade->type) {
		case VERMTYPE:
			verm_free((Verm*)grade);
			break;
	}
}

char *grade_to_string(Grade *grade)
{
	switch (grade->type) {
		case VERMTYPE:
			return verm_format((Verm *)grade);
		default:
			return NULL;
	}
}

void serialized_grade_free(SerializedGrade *grade)
{
	free(grade);
}

size_t serialized_grade_size_from_verm(void)
{
	size_t size = sizeof(uint32_t); // for type
	size += sizeof(uint8_t); // for verm->value content
	return size;
}

SerializedGrade *serialized_grade_from_verm(const Verm *verm, size_t *size)
{
	SerializedGrade	*grade;
	size_t	actual_size;
	size_t	expected_size;
	uint8_t	*ptr;

	expected_size = serialized_grade_size_from_verm();
	ptr = malloc(expected_size);
	grade = (SerializedGrade *)ptr;

	// TODO here is where flags could be added to ptr

	ptr += serialized_grade_buffer_write_verm(verm, ptr);

	actual_size = ptr - (uint8_t*)grade;

	assert(actual_size == expected_size);

	if (size)
		*size = expected_size;

        return grade;
}

Grade *grade_from_serialized(const SerializedGrade *serialized)
{
	return grade_from_serialized_grade_data((uint8_t *)serialized->data);
}

SerializedGrade *serialized_grade_from_grade(const Grade *grade, size_t *size)
{
	SerializedGrade *serialized = NULL;

	switch (grade->type) {
		case VERMTYPE:
			serialized = serialized_grade_from_verm((Verm *)grade, size);
	}

	return serialized;

}

Grade *grade_from_serialized_grade_data(uint8_t *buf)
{
	uint32_t type;

	type = serialized_grade_data_read_uint32_t(buf);

	switch (type) {
		case VERMTYPE:
			return (Grade *)verm_from_serialized_grade_data(buf, NULL);
		default:
			return NULL;
	}
}

Verm *verm_from_serialized_grade_data(const uint8_t *buf, size_t *size)
{
	const uint8_t *loc;
	Verm *verm;

	verm = verm_create(0);
	verm->type = VERMTYPE;

	loc = buf;
	loc += sizeof(uint32_t); // skip type
	verm_set_value(verm, serialized_grade_data_read_uint8_t(loc));
	loc += sizeof(uint8_t);

	if (size)
		*size = loc - buf;

	return verm;
}

uint32_t serialized_grade_data_read_uint32_t(const uint8_t *buf)
{
	return *((uint32_t*)buf);
}

uint8_t serialized_grade_data_read_uint8_t(const uint8_t *data)
{
	return *((uint8_t*)data);
}

size_t serialized_grade_buffer_write_verm(const Verm *verm, uint8_t *buf)
{
	uint8_t *loc;
	uint32_t type = VERMTYPE;
	uint8_t value = verm_get_value(verm);

	loc = buf;

	memcpy(loc, &type, sizeof(uint32_t));
	loc += sizeof(uint32_t);
	memcpy(loc, &value, sizeof(uint8_t));
	loc += sizeof(uint8_t);

	return loc - buf;
}
