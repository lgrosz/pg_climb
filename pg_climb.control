comment = 'rock climbing utilities'
default_version = '0.1'
module_pathname = '$libdir/pg_climb'
relocatable = true
