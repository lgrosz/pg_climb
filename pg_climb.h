#include <stddef.h>
#include <stdint.h>

#ifndef PG_CLIMB_H

// Grade Types - these are defines to avoid confusion about what type the
// compiler decides an enum to be
#define ANYTYPE	0
#define VERMTYPE	1
#define FONTTYPE	2
#define YDSTYPE	3

// Data Structures
typedef struct {
	void *data;
	uint32_t type;
} Grade;

typedef struct {
	uint8_t *value;
	uint32_t type; /* GRADE_TYPE_VERM */
} Verm;

typedef struct {
	uint8_t *value;
	uint32_t type; /* GRADE_TYPE_FONT */
} Font;

typedef struct {
	uint8_t *value;
	uint32_t type; /* GRADE_TYPE_YDS */
} Yds;

// This is a serialized grade. Some header data and flags can be added to the
// structure to determine how it should be interpreted, but for now it's just
// data. Here is how the data should be formatted.
//
// Open-ended, discrete scales (like the V-scale and Font-scale) are just
// represented by a single integer despite appearinging like they contain more
// than one component.
//
// <verm-type>
// [uint8_t]
//
// <font-type>
// [uint8_t]
//
// <yds-type>
// [uint8_t]
typedef struct {
	char data[1];
} SerializedGrade;

// Type Functions
const char *grade_type_name(uint32_t type);
uint32_t grade_type_from_typmod(const char *);
int typmod_string(char **, int32_t typmod);

// Verm Functions
Verm *verm_create(uint8_t initial_value);
void verm_free(Verm *verm);
uint8_t verm_get_value(const Verm *verm);
void verm_set_value(Verm *verm, uint8_t value);
int verm_parse(Verm *verm, const char *str);
Verm *verm_from_string(const char *str);
char *verm_format(const Verm *verm);

// Font Functions
Font *font_create(uint8_t initial_value);
void font_free(Font *font);
uint8_t font_get_value(const Font *font);
void font_set_value(Font *font, uint8_t value);
int font_parse(Font *font, const char *str);
Font *font_from_string(const char *str);
char *font_format(const Font *font);

// YDS Functions
Yds *yds_create(uint8_t initial_value);
void yds_free(Yds *yds);
uint8_t yds_get_value(const Yds *yds);
void yds_set_value(Yds *yds, uint8_t value);
int yds_parse(Yds *yds, const char *str);
Yds *yds_from_string(const char *str);
char *yds_format(const Yds *yds);

// Grade Functions
Grade *grade_from_string(const char *str, uint32_t type_hint);
void grade_free(Grade *grade);
char *grade_to_string(Grade *grade);

// Serialization Functions
void serialized_grade_free(SerializedGrade *grade);
size_t serialized_grade_size_from_verm(void);
SerializedGrade *serialized_grade_from_verm(const Verm *verm, size_t *size);
size_t serialized_grade_size_from_font(void);
SerializedGrade *serialized_grade_from_font(const Font *font, size_t *size);
size_t serialized_grade_size_from_yds(void);
SerializedGrade *serialized_grade_from_yds(const Yds *yds, size_t *size);
SerializedGrade *serialized_grade_from_grade(const Grade *grade, size_t *size);
Grade *grade_from_serialized(const SerializedGrade *serialized);

// Serialization Buffer Functions
Grade *grade_from_serialized_grade_data(uint8_t *buf);
Verm *verm_from_serialized_grade_data(const uint8_t *buf, size_t *size);
Font *font_from_serialized_grade_data(const uint8_t *buf, size_t *size);
Yds *yds_from_serialized_grade_data(const uint8_t *buf, size_t *size);
uint32_t serialized_grade_data_read_uint32_t(const uint8_t *buf);
uint8_t serialized_grade_data_read_uint8_t(const uint8_t *data);
size_t serialized_grade_buffer_write_verm(const Verm *verm, uint8_t *buf);
size_t serialized_grade_buffer_write_font(const Font *font, uint8_t *buf);
size_t serialized_grade_buffer_write_yds(const Yds *yds, uint8_t *buf);

#endif
