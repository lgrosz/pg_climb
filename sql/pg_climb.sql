CREATE EXTENSION pg_climb version '0.1';
 
SELECT EXISTS (
    SELECT 1 FROM pg_type WHERE typname = 'grade'
);

SELECT grade_out('V5'::grade);

-- invalid grade because only one type can be specified
CREATE TABLE grades_1(grade grade(verm,font));

-- invalid because "nope" isn't a grade type
CREATE TABLE grades_2(grade grade(nope));

