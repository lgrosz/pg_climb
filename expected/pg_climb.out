CREATE EXTENSION pg_climb version '0.1';
 
SELECT EXISTS (
    SELECT 1 FROM pg_type WHERE typname = 'grade'
);
 exists 
--------
 t
(1 row)

SELECT grade_out('V5'::grade);
 grade_out 
-----------
 V5
(1 row)

-- invalid grade because only one type can be specified
CREATE TABLE grades_1(grade grade(verm,font));
ERROR:  typmod array must contain exactly one value
LINE 1: CREATE TABLE grades_1(grade grade(verm,font));
                                    ^
-- invalid because "nope" isn't a grade type
CREATE TABLE grades_2(grade grade(nope));
ERROR:  parameter value not a valid typmod
LINE 1: CREATE TABLE grades_2(grade grade(nope));
                                    ^
